### ISSUE TITLE
Provide seed to DenseReparameterization layer

### ISSUE THREAD

**Issue by danijar** [tensorflow/probability#115]
I use DenseReparameterization for the transition function of a simple state-space model. To sample posterior sequences, I need to auto-regressively apply the layer to its own output inside of a symbolic loop. However, the samples are not consistent because the layer uses a different weight matrix at each iteration of the loop. Is it possible or are there plans to pass the seed into `layer(inputs, seed=0)` in order to sample consistent sequences?

**Comment by dustinvtran** [comment]
That's a good question. We mimicked the randomness approach from Keras' dropout layer. Do you know what their recommendation is?

**Comment by danijar** [comment]
I don't know. @fchollet?

**Comment by danijar** [comment]
A simple API for this would be to add the seed as optional argument to `__call__`.

**Comment by dustinvtran** [comment]
It looks like the `seed` is in the __init__ of tf.keras.layers.Dropout (https://www.tensorflow.org/api_docs/python/tf/layers/Dropout).

That should also work for us. Carry around the seed, and generate a new per-op seed at each random op call. (See [`tfp.distributions.SeedStream`](https://github.com/tensorflow/probability/blob/r0.3/tensorflow_probability/python/distributions/seed_stream.py)). Contributions welcome.

**Comment by dustinvtran** [comment]
@srvasude 

On second thought, this is actually difficult with the current design because any randomness is imposed by the user via the `{kernel,bias}_posterior_tensor_fn` arguments. Perhaps this is another example of why we need to redesign the API to avoid these generic callables.

**Comment by danijar** [comment]
How would passing the seed into `__init__` solve this? If I need to generate multiple sequences using the same Bayesian transition model, I'd have to reuse the weights but not the seed within the same session run.

**Comment by dustinvtran** [comment]
By passing a seed into `__init__`, you can initialize a seed stream which lets you use a different seed at each call().

**Comment by danijar** [comment]
But I couldn't use it to use the same seed for multiple calls? The whole point of this request was that I can sample a consistent autoregressive sequence `p(x_1:T,theta)~p(theta)p(x_1)prod_t=2..T p(x_t|x_t-1,theta)`. I want to implement `int p(x_t|x_t-1,theta)q(theta) dtheta` as a probabilistic layer, but the seed for the parameters needs to stay fixed over all time steps when sampling a sequence.

**Comment by dustinvtran** [comment]
As I mentioned, you can use the exact same seed by passing `{kernel,bias}_posterior_tensor_fn` as `lambda d: d.sample(seed=exact_same_seed)`.

By passing a seed into `__init__` (and a necessary redesign of the sampling op's API), you can also use a different seed at each call(), but fortunately still deterministically determine it via seed_stream.

In other words, both options are possible.

**Comment by dustinvtran** [comment]
> it would not be possible to generate multiple sequences from my Bayesian state space model within the same session run,

Yeah you can't do that unless you add `seed` to call; or otherwise you build a new layer and set its variables to be the previous layers.

I'm still not sure passing `{kernel,bias}_posterior_tensor_fn` during `__init__` is the right way to instantiate tensors given a distribution. Any ideas on redesigning that are welcome—whether it be during the init or the call.

**Comment by danijar** [comment]
Can you remind me which use cases are currently covered by `kernel_posterior_tensor_fn`? It almost seems to me that in a reparameterization layer, you always wanted to just do `dist.sample()`. If you wanted multiple samples for a better estimate, you'd need to implement a new layer anyway, right? To me the straightforward change would be to allow passing the seed to `__call__`. If it's not set, it defaults to `None` and uses an internal seed stream of the layer.
